### PR TITLE
Singularity fixes

### PR DESCRIPTION
**Pull request by metalgearsloth**
* Fix exception for visualizer state changes
* Tiles consumed properly (needs to be checked every n seconds as StartCollide is only run once)
* Range is an actual sensible value (based on SS13 value) so it shouldn't lag the heck.
* Non-hard so it doesn't bounce off of everything
* Consumption fixed again (layer problem).
* Deletions use both collision and lookup so all non-physics stuff gets counted *but* physics stuff is also deleted when colliding just in case.

Fixes: https://github.com/space-wizards/space-station-14/issues/4381

:cl:
- fix: Singulo slightly less janky.

### CODE REVIEW

**Comment by metalgearsloth** [comment]
Containment fields + emitters are still probably broken but that's for the next PR.

**Comment by metalgearsloth** [comment]
Okay so the problem is the client needs knowledge of how collision works for singulo so we can get nice hard collisions with containment and no jank mispredicts for ghosts but the problem is containment is all server-side so that requires a lot more work so now it's definitely next PR.

File: Content.Client/Singularity/Components/ContainmentFieldComponent.cs [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
 {
-    public class ContainmentFieldComponent : Component
+    [RegisterComponent]
+    [ComponentReference(typeof(SharedContainmentFieldComponent))]
+    public class ContainmentFieldComponent : SharedContainmentFieldComponent
     {
-        public override string Name => "ContainmentField";
-
+        // Jesus what is this code.
+        // Singulo cleanup WHEEENNN
```
Comment (GraniteSidewalk) [inline Content.Client/Singularity/Components/ContainmentFieldComponent.cs:13]: :(

**Comment by metalgearsloth** [comment]
NVM fixed the collision stuff this PR. Should work decently now (at least for destroying the station) as it is contained correctly and will delete the station more or less correctly.

File: Content.Server/Physics/Controllers/SingularityController.cs [hunk 2]
```diff
@@ -1,106 +1,56 @@
...
-
-            foreach (var grid in _mapManager.FindGridsIntersecting(component.Owner.Transform.MapID, worldBox))
-            {
-                foreach (var tile in grid.GetTilesIntersecting(worldBox))
-                {
-                    grid.SetTile(tile.GridIndices, Tile.Empty);
-                }
-            }
+            physics.BodyStatus = BodyStatus.InAir;
+            physics.ApplyLinearImpulse(pushVector.Normalized + (1f / singularity.Level) * physics.Mass);
```
Comment (GraniteSidewalk) [inline Content.Server/Physics/Controllers/SingularityController.cs:52]: No resetting inertia after each tick? That would be more true to SS13 singulos but I don't know if you want that.

File: Content.Server/Singularity/Components/ServerSingularityComponent.cs [hunk 3]
```diff
@@ -60,6 +60,8 @@ Level switch
                 _ => 0
             };
 
+        public float MoveAccumulator;
```
Comment (GraniteSidewalk) [inline Content.Server/Singularity/Components/ServerSingularityComponent.cs:63]: Unused?

File: Content.Server/Singularity/EntitySystems/SingularitySystem.cs [hunk 4]
```diff
@@ -21,56 +37,157 @@ public override void Initialize()
...
+        }
+
+        private float DestroyTileRange(ServerSingularityComponent component)
+        {
+            return component.Level - 0.5f;
+        }
+
+        private void HandleDestroy(ServerSingularityComponent component, IEntity entity)
+        {
+            // TODO: Need singuloimmune tag
```
Comment (GraniteSidewalk) [inline Content.Server/Singularity/EntitySystems/SingularitySystem.cs:103]: Being singulo-eatable should probably be opt-in, not opt-out (just a note)

File: Content.Server/Singularity/EntitySystems/SingularitySystem.cs [hunk 5]
```diff
@@ -21,56 +37,157 @@ public override void Initialize()
...
+            var destroyRange = DestroyTileRange(component);
+
+            foreach (var entity in _lookup.GetEntitiesInRange(component.Owner.Transform.MapID, worldPos, pullRange))
+            {
+                // I tried having it so level 6 can de-anchor. BAD IDEA, MASSIVE LAG.
+                if (entity == component.Owner ||
+                    !entity.TryGetComponent<PhysicsComponent>(out var collidableComponent) ||
+                    collidableComponent.BodyType == BodyType.Static) continue;
+
+                if (entity.HasComponent<GhostComponent>() ||
```
Comment (GraniteSidewalk) [inline Content.Server/Singularity/EntitySystems/SingularitySystem.cs:158]: Even if we aren't adding SinguloEatableComponent yet you can probably condense this into a function for easy editing.

File: Content.Server/Singularity/EntitySystems/SingularitySystem.cs [hunk 6]
```diff
@@ -21,56 +37,157 @@ public override void Initialize()
...
+                    entity.HasComponent<IMapGridComponent>() ||
+                    entity.HasComponent<MapComponent>() ||
+                    entity.IsInContainer()) continue;
+
+                var vec = worldPos - entity.Transform.WorldPosition;
+
+                if (vec.Length < destroyRange - 0.01f) continue;
+
+                var speed = vec.Length * component.Level * 10 * collidableComponent.Mass / 10;
```
Comment (GraniteSidewalk) [inline Content.Server/Singularity/EntitySystems/SingularitySystem.cs:167]: * 10 / 10

Comment (GraniteSidewalk) [inline Content.Server/Singularity/EntitySystems/SingularitySystem.cs:167]: \*10 / 10

File: Content.Server/Singularity/Components/ServerSingularityComponent.cs (hunk 3 above)
Comment (metalgearsloth) [reply Content.Server/Singularity/Components/ServerSingularityComponent.cs:63]: It's used on the controller.

File: Content.Server/Singularity/EntitySystems/SingularitySystem.cs (hunk 4 above)
Comment (metalgearsloth) [reply Content.Server/Singularity/EntitySystems/SingularitySystem.cs:103]: The fact that 99% of entities are destroyed by the singulo I think it'd be more annoying to make it opt-in, especially when you'd have to add a tagcomponent to effectively everything.

File: Content.Server/Singularity/EntitySystems/SingularitySystem.cs (hunk 5 above)
Comment (metalgearsloth) [reply Content.Server/Singularity/EntitySystems/SingularitySystem.cs:158]: I guess, even if it is only used in 1 spot.

File: Content.Server/Singularity/EntitySystems/SingularitySystem.cs (hunk 6 above)
Comment (metalgearsloth) [reply Content.Server/Singularity/EntitySystems/SingularitySystem.cs:167]: Those values need to be changed anyway for balance reasons but sure.

**Pull request merged**